libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,937 @@
+// Copyright 2019-2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(ia0): Remove when the module is used.
+#![allow(dead_code)]
+
+use crate::bitfield::*;
+use crate::{Storage, StorageIndex, StoreError, StoreResult};
+use alloc::vec::Vec;
+use core::cmp::min;
+
+type WORD = u32;
+
+/// Size of a word in bytes.
+///
+/// Currently, the store only supports storages where a word is 4 bytes.
+const WORD_SIZE: usize = core::mem::size_of::<WORD>();
+
+/// Maximum size of a page in bytes.
+///
+/// Currently, the store only supports storages where pages are between 8 and 1024 [words].
+///
+/// [words]: constant.WORD_SIZE.html
+const MAX_PAGE_SIZE: usize = 4096;
+
+/// Maximum number of erase cycles.
+///
+/// Currently, the store only supports storages where the maximum number of erase cycles fits on 16
+/// bits.
+const MAX_ERASE_CYCLE: usize = 65535;
+
+/// Maximum page index.
+///
+/// Thus the maximum number of pages is one more than this number. Currently, the store only
+/// supports storages where the number of pages is between 3 and 64.
+const MAX_PAGE_INDEX: usize = 63;
+
+/// Maximum key index.
+///
+/// Thus the number of keys is one more than this number. Currently, the store only supports 4096
+/// keys.
+const MAX_KEY_INDEX: usize = 4095;
+
+/// Maximum length in bytes of a user payload.
+///
+/// Currently, the store only supports values smaller than 1024 bytes.
+const MAX_VALUE_LEN: usize = 1023;
+
+/// Maximum number of updates per transaction.
+///
+/// Currently, the store only supports transactions with at most 31 updates.
+const MAX_UPDATES: usize = 31;
+
+/// Maximum number of words per virtual page.
+const MAX_VIRT_PAGE_SIZE: usize = div_ceil(MAX_PAGE_SIZE, WORD_SIZE) - CONTENT_WORD;
+
+/// Word with all bits set to one.
+const ERASED_WORD: u32 = 0xffffffff;
```

Comment:
Yes it's possible. I changed it to `!(0 as WORD)` which represents better the fact that it's a word with all bits set to 1. It happens to also be the maximum value but it's one more step of reasoning and looks at this value as a number instead of a sequence of bits, which breaks semantics (`WORD` means sequence of bits of the smallest unit of flash, `usize` means mathematical number, i.e. no overflow).